File: 2022/swift/Sources/advent2022/Day2.swift
struct Day2 : Day {
    let sampleInput = ["A Y", "B X", "C Z"];

    enum ParseError : Error {
        case InvalidToken (_ text :Substring)
    }
    typealias Parser<A> = (Substring) throws -> A

    enum Play :Int, CaseIterable {
        case Rock = 1, Paper = 2, Scissors = 3
        func beats (_ p :Play) -> Bool { Play.allCases[self.rawValue % Play.allCases.count] == p }
        func score (_ p :Play) -> Int { self.beats(p) ? 6 : p.beats(self) ? 0 : 3 }
    }
    let toPlay :Parser<Play> = {
        switch $0 {
        case "A", "X": return .Rock
        case "B", "Y": return .Paper
        case "C", "Z": return .Scissors
        default: throw ParseError.InvalidToken($0)
        }
    }

    enum Action {
        case Win, Lose, Draw
    }
    let toAction :Parser<Action> = {
        switch $0 {
            case "X": return .Lose
            case "Y": return .Draw
            case "Z": return .Win
            default: throw ParseError.InvalidToken($0)
        }
    }

    func readStrategy<A, B> (_ input :[String], _ cvt1 :Parser<A>, _ cvt2 :Parser<B>) throws -> [(A, B)] {
        try input.map({ $0.split(separator: " ") }).map({ (try cvt1($0[0]), try cvt2($0[1])) })
    }

    func score1 (_ p1 :Play, _ p2 :Play) -> Int { p1.score(p2) + p2.rawValue }
    func part1 () throws -> String { String(try readStrategy(readInput(2), toPlay, toPlay).map(score1).reduce(0, +)) }

    func pickAction (_ p1 :Play, _ a2 :Action) -> Play {
        switch a2 {
        case .Win: return Play.allCases[p1.rawValue % Play.allCases.count]
        case .Lose: return Play.allCases[(p1.rawValue + 1) % Play.allCases.count]
        case .Draw: return p1
        }
    }
    func score2 (_ p1 :Play, a2 :Action) -> Int { score1(p1, pickAction(p1, a2)) }
    func part2 () throws -> String { String(try readStrategy(readInput(2), toPlay, toAction).map(score2).reduce(0, +)) }
}

File: 2022/scala/src/main/scala/Day2.scala
object Day2 extends Day(2):

  enum Play:
    case Rock, Paper, Scissors

    def stronger = Play.values((ordinal + 1) % Play.values.length)
    def weaker = Play.values((ordinal + 2) % Play.values.length)
    def beats (p :Play) = stronger == p
    def score (p :Play) = if this.beats(p) then 6 else if p.beats(this) then 0 else 3

  import Play._
  object Play:
    def from (str :String) = str match
      case "A" | "X" => Rock
      case "B" | "Y" => Paper
      case "C" | "Z" => Scissors

  enum Action:
    case Win, Lose, Draw

  import Action._
  object Action:
    def from (str :String) = str match
      case "X" => Lose
      case "Y" => Draw
      case "Z" => Win

  def readStrategy[A, B] (input :Seq[String], cvt1 :String => A, cvt2 :String => B) =
    input.map(_.split(" ")).map(pp => (cvt1(pp(0)), cvt2(pp(1))))

  def score1 (p1 :Play, p2 :Play) = p1.score(p2) + p2.ordinal+1

  override def answer1 (input :Seq[String]) =
    readStrategy(input, Play.from, Play.from).map(score1).sum.toString

  def pickAction (p1 :Play, a2 :Action) = a2 match
    case Win => p1.stronger
    case Lose => p1.weaker
    case Draw => p1

  def score2 ( p1 :Play, a2 :Action) = score1(p1, pickAction(p1, a2))

  override def answer2 (input :Seq[String]) =
    readStrategy(input, Play.from, Action.from).map(score2).sum.toString
